src/clp_ffi_js/ir/StreamReader.cpp

```diff
@@ -99,63 +99,107 @@ auto StreamReader::get_num_events_buffered() const -> size_t {
     return m_encoded_log_events.size();
 }
 
-auto StreamReader::deserialize_range(size_t begin_idx, size_t end_idx) -> size_t {
-    constexpr size_t cFullRangeEndIdx{0};
-    if (0 != begin_idx || cFullRangeEndIdx != end_idx) {
-        throw ClpFfiJsException{
-                clp::ErrorCode::ErrorCode_Unsupported,
-                __FILENAME__,
-                __LINE__,
-                "Partial range deserialization is not yet supported."
-        };
+auto StreamReader::get_filtered_log_event_map() const -> FilteredLogEventMapType {
+    if (std::nullopt == m_filtered_log_event_map) {
+        return FilteredLogEventMapType(emscripten::val::null());
     }
+
+    return FilteredLogEventMapType(emscripten::val::array(m_filtered_log_event_map.value()));
+}
+
+auto StreamReader::build() -> size_t {
     if (nullptr != m_stream_reader_data_context) {
         constexpr size_t cDefaultNumReservedLogEvents{500'000};
         m_encoded_log_events.reserve(cDefaultNumReservedLogEvents);
+
+        std::string logtype;
+        constexpr size_t cDefaultReservedMessageLength{512};
+        logtype.reserve(cDefaultReservedMessageLength);
         while (true) {
             auto result{m_stream_reader_data_context->get_deserializer().deserialize_log_event()};
-            if (false == result.has_error()) {
-                m_encoded_log_events.emplace_back(std::move(result.value()));
-                continue;
-            }
-            auto const error{result.error()};
-            if (std::errc::no_message_available == error) {
-                break;
+            if (result.has_error()) {
+                auto const error{result.error()};
+                if (std::errc::no_message_available == error) {
+                    break;
+                }
+                if (std::errc::result_out_of_range == error) {
+                    SPDLOG_ERROR("File contains an incomplete IR stream");
+                    break;
+                }
+                throw ClpFfiJsException{
+                        clp::ErrorCode::ErrorCode_Corrupt,
+                        __FILENAME__,
+                        __LINE__,
+                        "Failed to deserialize: "s + error.category().name() + ":" + error.message()
+                };
             }
-            if (std::errc::result_out_of_range == error) {
-                SPDLOG_ERROR("File contains an incomplete IR stream");
-                break;
+            auto const& log_event = result.value();
+            auto const& message = log_event.get_message();
+
+            logtype.clear();
+            logtype = message.get_logtype();
```

Comment:
Why clear before copy?

Reply:
Yeah, the existing code is a bit strange. I thought we had changed it, but I guess not. Your suggestion is how I would write it.